plugins/pantheon-files-cloud/plugin.vala

```diff
@@ -0,0 +1,127 @@
+/***
+    Copyright (c) 2018 elementary LLC <https://elementary.io>
+
+    Pantheon Files is free software; you can redistribute it and/or
+    modify it under the terms of the GNU General Public License as
+    published by the Free Software Foundation; either version 2 of the
+    License, or (at your option) any later version.
+
+    Pantheon Files is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+    General Public License for more details.
+
+    You should have received a copy of the GNU General Public License along
+    with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+    Author(s):  Fernando da Silva Sousa <wild.nando@gmail.com>
+***/
+
+public class Marlin.Plugins.Cloud.Plugin : Marlin.Plugins.Base {
+    Marlin.AbstractSidebar? sidebar;
+    CloudProviders.Collector collector;
+    CloudProviders.Provider[] providers_connected;
+
+    public Plugin () {
+        providers_connected = {};
+        collector = CloudProviders.Collector.dup_singleton ();
+        collector.providers_changed.connect (on_providers_changes);
+    }
+
+    /**
+     * Assign loaded sidebar to plugin's sidebar reference
+     *
+     * @param a instance of Marlin.AbstractSidebar
+     */
+    public override void sidebar_loaded (Gtk.Widget widget) {
+        sidebar = widget as Marlin.AbstractSidebar;
+        on_providers_changes ();
+    }
+
+    /**
+     * Plugin hook that triggers when sidebar receives a update request on
+     * Marlin's code
+     *
+     * @param a instance of Marlin.AbstractSidebar
+     */
+    public override void update_sidebar (Gtk.Widget widget) {
+        foreach (var provider in collector.get_providers ()) {
+            foreach (var account in provider.get_accounts ()) {
+                add_account_to_sidebar (account, provider);
+            }
+        }
+    }
+
+    void on_providers_changes () {
+        //  Listen to new accounts
+        unowned GLib.List<CloudProviders.Provider> providers = collector.get_providers ();
+        foreach (var provider in providers) {
+            //  Avoid listening to same provider again
+            if (!(provider in providers_connected)) {
+                providers_connected += provider;
```

Comment:
What happens if a provider is removed? From what I see, current code doesn't account for that at all. Please correct me if I'm wrong.

Reply:
Ending the client process causes the bookmark to be removed from Sidebar without problems.

Reply:
In which way and where in the code the bookmark is removed from the sidebar? I don't see it. `providers_connected` will always grow in size, because nothing is ever removed from it.

Reply:
Added code to remove lost providers from list.